### PR TITLE
Added motoman_mh5 IKfast plugin

### PR DESCRIPTION
**Pull request by nsnitish**

### CODE REVIEW

**Comment by nsnitish** [comment]
@shaun-edwards! Please have a look

**Comment by shaun-edwards** [comment]
@nsnitish,  The failure occurs in the indigo build.  The problem is an ambiguous call to `isnan`.  The easiest solution is to add the `std::` namespace to the call of `isnan` in `*_ikfast_solver, as was done [here](https://github.com/uos/katana_driver/commit/875324a96c9be1d221b4db951c4df8a1c3ff0c4e).

This would require changing an auto-generated ikfast file (not desirable).

@gavanderhoorn solved this problem, this [way](https://github.com/ros-industrial/fanuc/commit/c8eb156d14fb96b932c90a19e61181a0e04b465a).  However, when I use this, I get the following error:
```
motoman_mh5_manipulator_ikfast_solver.cpp:1631:77: error: ‘>>’ should be ‘> >’ within a nested template argument list
                                                                       IkReal>>

<snip>

motoman_mh5_manipulator_ikfast_moveit_plugin.cpp:1222:36: error: ‘>>’ should be ‘> >’ within a nested template argument list
   std::vector<IkSolutionList<IkReal>> solution_set;
```
This doesn't occur in the fanuc package, because it's `*_ikfast_solver.cpp` and `*_moveit_plugin.cpp` do not have examples of `>>`.  I'm not sure why this is.

Unless @gavanderhoorn provides some insight, I suggest adding `std::` to each `isnan` call.

**Comment by gavanderhoorn** [comment]
@shaun-edwards wrote:
> This doesn't occur in the fanuc package, because it's `*_ikfast_solver.cpp` and `*_moveit_plugin.cpp` do not have examples of `>>`.  I'm not sure why this is.

The IK plugins in `fanuc` are rather old (they're overdue for some updates: ros-industrial/fanuc#123), so the `>>` and `isnan(..)` stuff was probably added to the IKFast template after they were generated.

**Comment by nsnitish** [comment]
@shaun-edwards! Can you have a look?

File: motoman_mh5_ikfast_manipulator_plugin/package.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?xml version='1.0' encoding='ASCII'?>
```
Comment (shaun-edwards) [inline motoman_mh5_ikfast_manipulator_plugin/package.xml:1]: Please clean up/delete the comments in this file.

File: motoman_mh5_ikfast_manipulator_plugin/package.xml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+<package>
+  <name>motoman_mh5_ikfast_manipulator_plugin</name>
+  <version>0.0.0</version>
+  <description>The motoman_mh5_ikfast_manipulator_plugin package</description>
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="nitishsharma@todo.todo">nitishsharma</maintainer>
```
Comment (shaun-edwards) [inline motoman_mh5_ikfast_manipulator_plugin/package.xml]: Add your PlusOne email here.

File: motoman_mh5_ikfast_manipulator_plugin/package.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
```
Comment (shaun-edwards) [inline motoman_mh5_ikfast_manipulator_plugin/package.xml]: Set the license to Apache 2.0

**Review by shaun-edwards — CHANGES_REQUESTED** [review]
Small changes requested.  Then it looks like we are good to go.

**Comment by nsnitish** [comment]
@shaun-edwards! Are there any other changes required?

**Review by shaun-edwards — APPROVED** [review]
Looks good.  Thanks for the contribution!

**Comment by gavanderhoorn** [comment]
Just curious: the generated `motoman_mh5_manipulator_ikfast_solver.cpp` appears to be 4.3MB (!) in size. That seems a tad excessive.

@nsnitish: can you describe how you generated this solver?

**Comment by gavanderhoorn** [comment]
Two other comments:

 - `update_ikfast_plugin.sh` contains references to paths local to @nsnitish' machine (`/home/nitishsharma/..`)
 - the introduced pkg does not follow the layout and formatting of other MoveIt IKFast plugins we have (in the ABB and Fanuc repositories fi). Would we want to harmonize across all ros-i repositories and update this one?

**Comment by nsnitish** [comment]
@gavanderhoorn! I followed [this tutorial](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/ikfast_tutorial.html).

**Comment by gavanderhoorn** [comment]
And you installed OpenRAVE from source? Can you tell me which commit your `openrave` clone is at? What version of `sympy` did you use (`dpkg -l | grep sympy`)?

Which links did you select for `baselink` and `eelink`?

I've just run the same command, but using the `personalrobotics/ros-openrave` Docker image, and the generated IK solver is 103KB. That is quite a significant difference.

Running the iktests (with a different generator command) I get the following output:
```
testik, success rate: 1.000000, wrong solutions: 0.000000, no solutions: 0.000000, missing solution: 0.455000
```

That is not really good: basically the solver can only find a solution for about 50% of the queries (but if it finds one, they are all ok).

Have you ran the OpenRAVE iktests?

**Comment by nsnitish** [comment]
@gavanderhoorn! Thanks for pointing this out. It's kinda weird but the problem is in the [clang formatting](http://moveit.ros.org/documentation/contributing/code/) that I used. Auto-generated file is 571 KB.

I haven't ran the OpenRAVE iktests but I have tested the plugin with [moveit_kinematics_tests](https://github.com/ros-planning/moveit_kinematics_tests).

**Comment by shaun-edwards** [comment]
It's odd that ikfast wouldn't be better at solving the mh5 kinematics.  It is a standard industrial manipulator with a spherical wrist.  If ikfast has trouble with this robot, then I would expect it to have trouble with a lot of other robots we have.

**Comment by gavanderhoorn** [comment]
@nsnitish wrote:
> @gavanderhoorn! Thanks for pointing this out. It's kinda weird but the problem is in the [clang formatting](http://moveit.ros.org/documentation/contributing/code/) that I used. Auto-generated file is 571 KB.

So that would mean that the formatting introduced about 3.8MB of additional characters?

> I haven't ran the OpenRAVE iktests but I have tested the plugin with [moveit_kinematics_tests](https://github.com/ros-planning/moveit_kinematics_tests).

And that all succeeded?

It seems strange that OpenRAVE's own tests would indicate this low a success rate, and then 'our' test does succeed. It's not impossible though, as I've seen other people claim that `iktests` failed horribly for them, but the solver worked 'alright'.

But I think this deserves some more investigation.

@nsnitish: could you also answer my other questions please?

**Comment by gavanderhoorn** [comment]
@shaun-edwards:
> It's odd that ikfast wouldn't be better at solving the mh5 kinematics.  It is a standard industrial manipulator with a spherical wrist.  If ikfast has trouble with this robot, then I would expect it to have trouble with a lot of other robots we have.

I agree that it probably should be better, but I'm just reporting what I'm seeing.

If the generated solver is really not that good, we should at least mention that somewhere.

Related: ros-industrial/fanuc#99.

**Comment by nsnitish** [comment]
@gavanderhoorn! 
1. I am using the [latest stable commit of openRAVE](https://github.com/rdiankov/openrave/tree/latest_stable), sympy version 1.1. 

2. baselink is 0 and eelink is 6. 

3. Results from moveit_kinematics_tests for motoman_mh5
screenshot from 2017-07-10 19-10-00

**Comment by gavanderhoorn** [comment]
@nsnitish wrote:

> 1. I am using the [latest stable commit of openRAVE](https://github.com/rdiankov/openrave/tree/latest_stable), 

wow. That is quite old: [latest_stable...master](https://github.com/rdiankov/openrave/compare/latest_stable...master).

> sympy version 1.1. 

I'm surprised this works. But maybe there is such a difference between `latest-stable` and `master` that my previous experience with `sympy` doesn't apply here.

> baselink is 0 and eelink is 6.

ok.

> Results from moveit_kinematics_tests for motoman_mh5
screenshot from 2017-07-10 19-10-00

Ok. This doesn't tell me much but it looks like most tests are succeeding.

I still believe this deserves more investigation, as it's at least weird that OpenRAVE's own test claims that 50% of IK requests fail, while 'our' tests indicate that (almost) everything is fine.

**Comment by gavanderhoorn** [comment]
@nsnitish wrote:
> @gavanderhoorn! Thanks for pointing this out. It's kinda weird but the problem is in the [clang formatting](http://moveit.ros.org/documentation/contributing/code/) that I used. Auto-generated file is 571 KB.

this is unfortunate actually: we don't require clang-formatting in this repository, and it seems strange to do that to auto-generated code in the first place. But it's been merged, so lets not concern ourselves with that any more.

**Pull request merged**